frontend/src/App.test.tsx

```diff
@@ -475,3 +480,99 @@ describe("App.handleNewReport", () => {
     expect(SessionInfo.isSet()).toBe(true)
   })
 })
+
+// Using this to test the functionality provided through streamlit.experimental_set_query_params.
+// Please see https://github.com/streamlit/streamlit/issues/2887 for more context on this.
+describe("App.handlePageInfoChanged", () => {
+  beforeEach(() => {
+    // Reset the value of document.location.pathname.
+    window.history.pushState({}, "", "/")
+  })
+  it("does not override the pathname when setting query params", () => {
+    const wrapper = shallow(<App {...getProps()} />)
+    const app = wrapper.instance()
+    const pathname = "/foo/bar/"
+    // Set the value of document.location.pathname to pathname.
+    window.history.pushState({}, "", pathname)
+
+    const pageInfo = new PageInfo({
+      queryString: "flying=spaghetti&monster=omg",
+    })
+    const expectedUrl = `${pathname}?${pageInfo.queryString}`
+    const pushState = jest.spyOn(
```

Comment:
nit: Two small things here

* This can also be moved into the `beforeEach` like `wrapper` and `app`
* The name `pushStateSpy` would be a bit better here since it's more descriptive and also matches how spies are named in other tests